docs/buffer-metatable.md

```diff
@@ -0,0 +1,147 @@
+Any of these tables can be frozen before the call.
+
+When `__index` or `__newindex` is a function, VM will be allowed to ignore changes to the environment of those function.
+This is similar to function inlining functionality we have where an inlined function will no longer respect its environment.
+
+The freezing is performed to provide additional guarantees for field access and other metamethod evaluation to unlock potential optimization opportunities.
+This is similar to limitations placed on 'cdata' by luajit.
+Having said that, this RFC doesn't make a promise of a particular implementation making those optimizations, so should be viewed as a buffer object usability improvement first.
+
+Any chained metatables on table fields are not modified.
+This provides an option to have a more dynamic behavior, but giving up on potential performance improvements of a chained indexing access.
+This matches behavior of 'cdata' in luajit and will provide a familiarity for developers switching over.
+
+VM metatable lookups and `getmetatable` will look for the buffer object metatable first and then fall back to the global buffer metatable.
+This preserves the existing buffer extensibility point for hosts.
+
+`setmetatable` will still not be supported on buffer objects, the metatable reference is immutable.
+
+Equality checks in the VM will call `__eq` for buffers similar to tables and userdata.
+
+`__type` metatable key is ignored by `typeof`, just like it does for tables.
+As before, only host is allowed to define type names.
+This does not change how it behaved with a global host-provided metatable.
+
+Other metamethods (`__add`/`__len`/...) work as expected and do not change from how they work today for buffers with global host-provided metatable.
+
+Buffer structure size increase by the metatable aligns the data to a 16 byte boundary making it possible to store data which requires 16 byte alignment by the embedder.
+This will make buffer data match the alignment guarantee of Luau userdata objects.
+
+In order for the typechecker to understand buffers with attached metatables, we propose extending the intersections to be allowed on buffers, similar to `extern` types:
+
+`type Obj = buffer & { x: number, y: number, z: number }`
+
+## Forwards compatibility
+
+A potential evolution of the buffer type might move the data from the buffer memory block to a separate location with a redirection.
+
+Doing that will enable extending the buffer size without a major impact on performance (shrinking is a problem for range check eliminations).
+It might also allow buffer views and slices.
+
+In both cases, this change is compatible. The slice will reuse a section of the data but being a buffer it will have its own metatable.
+
+This makes it possible to have a large memory buffer and sub-allocate buffer objects with attached custom behaviors.
+
+## Drawbacks
```

Comment:
This doesn't seem to mention anything about the performance characteristics of `buffer` vs `table` (even if deserialized from buffers) wrt algorithms that does heavy reads/writes? If I have an algorithm that is very _heavy_ on reads/writes, it has to undergo a lot of overhead to turn a few bytes of a buffer into a `TValue` (or vice versa) whereas tables skips this ser/de.